query.go

```diff
@@ -120,190 +163,44 @@ func (s *ChainService) queryPeers(
-					// After timeout, we try to notify
-					// another of our peer goroutines to do
-					// a query until we get a signal to
-					// quit.
-					select {
-					case startQuery <- struct{}{}:
-					case <-quit:
-						return
-					case <-allQuit:
-						return
-					}
-
-					// At this point, we've sent
-					// startQuery.  We return if we've run
-					// through this section of code
-					// numRetries times.
-					if numRetries--; numRetries == 0 {
-						return
-					}
-
-				// After we're told to quit, we return.
-				case <-quit:
-					return
-
-				// After we're told to quit, we return.
-				case <-allQuit:
-					return
-
-				case <-startQuery:
-					// We're the lucky peer whose turn it
-					// is to try to answer the current
-					// query.
-					//
-					// If the sync peer hasn't tried yet
-					// and we aren't the sync peer, don't
-					// do anything but forward the message
-					// down the startQuery channel until
-					// the sync peer gets a shot.
-					//
-					// TODO: Add support for querying *all*
-					// peers simultaneously to avoid
-					// timeout delays.
-					if sp == syncPeer {
-						atomic.StoreUint32(
-							&syncPeerTries, 1)
-					}
-					if atomic.LoadUint32(&syncPeerTries) == 0 {
-						select {
-						case startQuery <- struct{}{}:
-						case <-quit:
-							return
-						case <-allQuit:
-							return
-						}
-						continue queryLoop
-					}
-
-					sp.subscribeRecvMsg(subscription)
-
-					// Don't want the peer hanging on send
-					// to the channel if we quit before
-					// reading the channel.
-					sentChan := make(chan struct{}, 1)
-					sp.QueueMessageWithEncoding(queryMsg,
-						sentChan, wire.WitnessEncoding)
-					select {
-					case <-sentChan:
-					case <-quit:
-						return
-					case <-allQuit:
-						return
-					}
-					timeout = time.After(qo.timeout)
-				default:
-				}
-			}
-		}(sp)
+		quitChan: subQuit,
 	}
 
-	// Kick off the query by sending the query totem into the startQuery
-	// channel.
-	startQuery <- struct{}{}
-
-	// This goroutine will wait until all of the peer-query goroutines have
-	// terminated, and then initiate a query shutdown.
-	go func() {
-		wg.Wait()
-
-		// If we timed out on each goroutine and didn't quit or time
-		// out on the main goroutine, make sure our main goroutine
-		// knows to quit.
-		select {
-		case <-allQuit:
-		default:
-			close(allQuit)
-		}
-
-		// Close the done channel, if any.
-		if qo.doneChan != nil {
-			close(qo.doneChan)
-		}
-
-		// Wait until all goroutines started by subscriptions have
-		// exited after we closed allQuit before letting the message
-		// channel get garbage collected.
-		subwg.Wait()
-	}()
-
 	// Loop for any messages sent to us via our subscription channel and
 	// check them for whether they satisfy the query. Break the loop if
 	// it's time to quit.
-	timeout := time.After(time.Duration(len(peers)+1) *
-		qo.timeout * time.Duration(qo.numRetries))
+	peerTimeout := time.Tick(qo.timeout)
```

Comment:
A `time.Ticker` should be used here, as otherwise, we can't actually shutdown the ticker and it'll go on forever with a goroutine leak.

Reply:
Thanks, will fix.